src/storage/storage_structure/lists/rel_update_store.cpp

```diff
@@ -0,0 +1,302 @@
+            }
+        }
+    } else {
+        for (auto direction : REL_DIRECTIONS) {
+            auto boundNodeID = direction == RelDirection::FWD ? srcNodeID : dstNodeID;
+            auto chunkIdx = StorageUtils::getListChunkIdx(boundNodeID.offset);
+            if (relUpdatesPerTablePerDirection[direction].contains(boundNodeID.tableID)) {
+                if (!relUpdatesPerTablePerDirection[direction]
+                         .at(boundNodeID.tableID)[chunkIdx]
+                         .contains(boundNodeID.offset)) {
+                    relUpdatesPerTablePerDirection[direction]
+                        .at(boundNodeID.tableID)[chunkIdx]
+                        .emplace(boundNodeID.offset,
+                            make_unique<RelUpdatesForNodeOffset>(relTableSchema));
+                }
+                relUpdatesPerTablePerDirection[direction]
+                    .at(boundNodeID.tableID)[chunkIdx]
+                    .at(boundNodeID.offset)
+                    ->deletedRelIDs.insert(relID);
+            }
+        }
+    }
+}
+
+uint64_t RelUpdateStore::getNumInsertedRelsForNodeOffset(
+    ListFileID& listFileID, node_offset_t nodeOffset) const {
+    auto relUpdatesForNodeOffset = getRelUpdatesForNodeOffsetIfExists(listFileID, nodeOffset);
+    if (relUpdatesForNodeOffset == nullptr) {
+        return 0;
+    }
+    return relUpdatesForNodeOffset->insertedRelsTupleIdxInFT.size();
+}
+
+void RelUpdateStore::readValues(
+    ListFileID& listFileID, ListHandle& listHandle, shared_ptr<ValueVector> valueVector) const {
+    auto numTuplesToRead = listHandle.getNumValuesToRead();
+    auto nodeOffset = listHandle.getBoundNodeOffset();
+    if (numTuplesToRead == 0) {
+        valueVector->state->initOriginalAndSelectedSize(0);
+        return;
+    }
+    auto vectorsToRead = vector<shared_ptr<ValueVector>>{valueVector};
+    auto columnsToRead = vector<ft_col_idx_t>{getColIdxInFT(listFileID)};
+    auto relNodeTableAndDir = getRelNodeTableAndDirFromListFileID(listFileID);
+    auto& listUpdates = relUpdatesPerTablePerDirection[relNodeTableAndDir.dir]
+                            .at(relNodeTableAndDir.srcNodeTableID)
+                            .at(StorageUtils::getListChunkIdx(nodeOffset))
+                            .at(nodeOffset);
+    insertedRels->lookup(vectorsToRead, columnsToRead, listUpdates->insertedRelsTupleIdxInFT,
+        listHandle.getStartElemOffset(), numTuplesToRead);
+    valueVector->state->originalSize = numTuplesToRead;
+}
+
+bool RelUpdateStore::hasAnyDeletedRelsInPersistentStore(
+    ListFileID& listFileID, node_offset_t nodeOffset) const {
+    auto relUpdatesForNodeOffset = getRelUpdatesForNodeOffsetIfExists(listFileID, nodeOffset);
+    if (relUpdatesForNodeOffset == nullptr) {
+        return false;
+    }
+    return !relUpdatesForNodeOffset->deletedRelIDs.empty();
+}
+
+void RelUpdateStore::updateRelIfNecessary(const shared_ptr<ValueVector>& srcNodeIDVector,
```

Comment:
I am having trouble following this code. You should add comments to your code to clarify the important steps. For example, I cannot see which parts of this code handles the case of updating a newly inserted rel.

Reply:
Added some comments in `updateRelIfNecessary()` function